### PR TITLE
GCS: Allow no-auth for testing purposes

### PR DESCRIPTION
**Pull request by snazy**
Although there is no "official" Google Cloud Storage emulator available yet, there is [one available](https://github.com/oittaa/gcp-storage-emulator) that allows at least some basic testing. To use an emulator, the client needs to be configured to use no authentication, otherwise it will fallback to "automatic credential detection".

### CODE REVIEW

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java [hunk 1]
```diff
@@ -18,13 +18,15 @@
  */
 package org.apache.iceberg.gcp;
 
+import com.google.api.client.util.Preconditions;
```
Comment (XN137) [inline gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:21]: wrong import?

File: gcp/src/main/java/org/apache/iceberg/gcp/gcs/GCSFileIO.java [hunk 2]
```diff
@@ -55,6 +56,7 @@
  * Overview</a>
  */
 public class GCSFileIO implements DelegateFileIO {
+
```
Comment (nastra) [inline gcp/src/main/java/org/apache/iceberg/gcp/gcs/GCSFileIO.java:59]: unnecessary change

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java [hunk 3]
```diff
@@ -18,13 +18,15 @@
...
 import java.io.Serializable;
 import java.util.Date;
 import java.util.Map;
 import java.util.Optional;
 import org.apache.iceberg.util.PropertyUtil;
 
 public class GCPProperties implements Serializable {
+
```
Comment (nastra) [inline gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:29]: unnecessary change

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java (hunk 1 above)
Comment (nastra) [reply gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:21]: yes, this is a wrong import and should use the one from `iceberg-guava`

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java [hunk 4]
```diff
@@ -90,6 +95,10 @@ public GCPProperties(Map<String, String> properties) {
       gcsOAuth2TokenExpiresAt =
           new Date(Long.parseLong(properties.get(GCS_OAUTH2_TOKEN_EXPIRES_AT)));
     }
+    gcsNoAuth = Boolean.parseBoolean(properties.getOrDefault(GCS_NO_AUTH, "false"));
+    Preconditions.checkState(
+        !(gcsOAuth2Token != null && gcsNoAuth),
+        "Must not configure " + GCS_NO_AUTH + " and " + GCS_OAUTH2_TOKEN);
```
Comment (nastra) [inline gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:101]:
```suggestion
        "Invalid auth settings: must not configure %s and %s", GCS_NO_AUTH, GCS_OAUTH2_TOKEN);
```

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java [hunk 5]
```diff
@@ -90,6 +95,10 @@ public GCPProperties(Map<String, String> properties) {
       gcsOAuth2TokenExpiresAt =
           new Date(Long.parseLong(properties.get(GCS_OAUTH2_TOKEN_EXPIRES_AT)));
     }
+    gcsNoAuth = Boolean.parseBoolean(properties.getOrDefault(GCS_NO_AUTH, "false"));
+    Preconditions.checkState(
```
Comment (nastra) [inline gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:99]: it would be good to also have a test that specifically checks that you can only set one of these

**Review by dimas-b — APPROVED** [review]

**Review by dimas-b — APPROVED** [review]

File: gcp/src/test/java/org/apache/iceberg/gcp/GCPPropertiesTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  public void testOAuthWithNoAuth() {
+    Assertions.assertThatIllegalStateException()
+        .isThrownBy(
+            () ->
+                new GCPProperties(ImmutableMap.of(GCS_OAUTH2_TOKEN, "oauth", GCS_NO_AUTH, "true")))
+        .withMessage(
+            String.format(
+                "Invalid auth settings: must not configure %s and %s",
+                GCS_NO_AUTH, GCS_OAUTH2_TOKEN));
+    Assertions.assertThatNoException()
```
Comment (nastra) [inline gcp/src/test/java/org/apache/iceberg/gcp/GCPPropertiesTest.java:40]: rather than just checking that no exception has been thrown, I think the better alternative would be:

```
    GCPProperties gcpProperties =
        new GCPProperties(ImmutableMap.of(GCS_OAUTH2_TOKEN, "oauth", GCS_NO_AUTH, "false"));
    assertThat(gcpProperties.noAuth()).isFalse();
    assertThat(gcpProperties.oauth2Token()).isPresent().get().isEqualTo("oauth");

    gcpProperties = new GCPProperties(ImmutableMap.of(GCS_NO_AUTH, "true"));
    assertThat(gcpProperties.noAuth()).isTrue();
    assertThat(gcpProperties.oauth2Token()).isNotPresent();
```

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java (hunk 3 above)
Comment (nastra) [reply gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:29]: @snazy can you please remove that newline change?

**Comment by nastra** [comment]
@bryanck could you review this one please?

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java (hunk 3 above)
Comment (snazy) [reply gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:29]: Nope - it's spottless that re-adds it everytime

Comment (nastra) [reply gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:29]: that's weird, it doesn't do that for me. Also the build would have failed if that code here wouldn't be spotless-compliant

**Comment by bryanck** [comment]
This looks good to me, as long as the format checks are passing.

**Review by bryanck — APPROVED** [review]

File: gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java (hunk 3 above)
Comment (nastra) [reply gcp/src/main/java/org/apache/iceberg/gcp/GCPProperties.java:29]: PR is ready to go once this diff is reverted

**Review by nastra — APPROVED** [review]
